sdlib/d/gc/thread.d

```diff
@@ -181,30 +181,32 @@ public:
 		import d.gc.hooks;
 		__sd_gc_pre_stop_the_world_hook();
 
-		stopTheWorldMutex.lock();
+		// Prevent any new threads from being created
+		createThreadLock.exclusiveLock();
 
 		uint count;
-		while (suspendRunningThreads(count++)
-			       || startingThreadCount.load() > 0) {
+
+		while (suspendRunningThreads(count++)) {
 			import sys.posix.sched;
 			sched_yield();
 		}
 	}
 
 	void restartTheWorld() shared {
+		// Allow thread creation again.
+		createThreadLock.exclusiveUnlock();
+
 		while (resumeSuspendedThreads()) {
 			import sys.posix.sched;
 			sched_yield();
 		}
 
-		stopTheWorldMutex.unlock();
-
 		import d.gc.hooks;
 		__sd_gc_post_restart_the_world_hook();
 	}
 
 	void scanSuspendedThreads(ScanDg scan) shared {
-		assert(stopTheWorldMutex.isHeld());
+		assert(createThreadLock.count == SharedLock.Exclusive);
```

Comment:
Probably want a method in SharedLock for this.

Reply:
Indeed. `isExcludivelyHeld` or something.

Reply:
I'm waiting on rwlock merge to make this change.